### ISSUE TITLE
Pytorch version?

### ISSUE THREAD

**Issue by shubbey** [thorben-frank/mlff#7]
Do you have any advice as for porting  this model to pytorch?  I have been experimenting with and developing other models in pytorch and it is being used (via jit) on the front-end of our program.  Your model is showing superior results from my previous cases (as far as I can tell without having trained it fully yet), so I would love to try it out in our program.  I don't think doing a native jax version on the front-end (C++) is feasible at the moment, although I could certainly use it for training if there was an easy way to take the output model and convert it into a format that libtorch could use.  At the moment I've been playing with porting your code file by file although this is quite tedious and probably error prone.   I've also considered modifying SpookyNet (our previous best model) since it is already in pytorch, although that presents its own issues.   I am just wondering if you have any suggestions.  Thank you.

**Comment by shubbey** [comment]
Just FYI,  I did some research and it seems that the best approach is to work with ONNX (I can convert both flax and torch models to that), so I'll see what it takes to integrate that on the front-end.  I will share any findings here, if anyone is interested.

**Comment by thorben-frank** [comment]
I did not know about ONNX, so I would be very interested in anything you experience with the tool on the way. 

I wanted to share a few thoughts and questions regarding porting the code to PyTorch. 

If I understood correctly, you already have the code for training your PyTorch spookynet model? 

I think most code of mlff are just helper functions for training, validation, running MD simulations and so on, such that the actual code required for so3krates should be only a minor part of the whole library. In case porting is still an option I’d be happy to pinpoint to the relevant parts of the code. 

I think as soon as ported to PyTorch it should be easy to integrate into your pipeline. Something I was wondering: How does your spookynet implementation handles the atomic neighbour lists? As you already encountered, in mlff atomic types (and neighbour lists) are padded to the largest structure, which gives a memory overheads when training on size diverse data sets. There is a more efficient way by effectively removing the batch axis and merge it with the neighbour lists axis. From a model perspective this shouldn’t be more than a few lines of extra code compared to the current implementation. Probably a well placed torch.scatter_add.

I think the potentially largest challenge in porting the code, is the heavy use of jax.vmap and flax.linen.vmap. I know that there have been some efforts to provide similar functionality in PyTorch using functorch but I don’t know what the current status is. If it is not working properly yet, one has to spend a little more thought on the additional axis in the tensors.

**Comment by shubbey** [comment]
I would prefer to use a native pytorch version if possible because I am more familar with pytorch and if this works out will likely need to modify the code to accommodate my other models.  For instance, we have used spookynet to predict chemical shifts for nmr simply by treating it as the "energy" term and making a few modifications.  I would like to try this with so3krates if it performs well for our energy/force model.

I did notice that spookynet handles batches differently (no need to pad) by merging.  Fortunately I didn't have to do anything special as this is fed into their compute_energy routine.  I posted an early version of my training routine for another user a while ago that shows basically what we do: https://github.com/OUnke/SpookyNet/issues/1#issuecomment-1079377906

I played with porting what I figured were the relevant parts of your model (mostly the stuff in layer/ if I recall, but it was still a fair amount and flax does handle things differently enough  than pytorch that it seemed very error prone.  Pytorch does have the functional stuff as you mentioned but I haven't used it enough to know where it stands.  I'd welcome any help or pointers in the right direction.  It seems to me that starting with SpookyNet and replacing the model part would be ideal. I wonder if it handles the input (spherical harmonics) in the same way?  Thank you!

**Comment by shubbey** [comment]
Hey, just wanted to bump this to see if you could give me some advice on a pytorch port.  I have been trying to get this working in onnx but it is an IT nightmare!  And I thought torch.jit was annoying.  

I finished training your results on 10% of my dataset and if I am reading them correctly, they are MUCH better than what I previously got with even my full dataset via SpookyNet. This doesn't seem possible so I likely have a units issue or something, but I am eager to confirm this.  It is a little bit challenging to make a direct comparison using the current tools although I will attempt that next.  As always, thanks for your help.  

[EDIT - 6/27] After running the same dataset on SpookyNet and correcting the units I'm now seeing comparable (albeit slightly worse) results, but it does seem to take longer to converge.  This is consistent with your paper.  I'm  still excited by your model since the paper indicates it is smaller and hence should be faster in production, and once my current run finishes I'll be able to see how the accuracy compares.

**Comment by shubbey** [comment]
Some preliminary stats: I ran a test on about 100k molecules from my dataset, using fixed train/valid/test splits and the same MSE loss function of ```0.01*ΔEnergy/molecule + 0.99*ΔForce/component```.  I compared out-of-the-box so3krates to  my spookynet implementation along roughly the same metric, which was a decay of the LR from 1e-3 to 1-5.  We have different decay functions so it's not super scientific, but these seem like decent stopping metrics.  Both took approximately 1000 epochs or 1M steps (1 epoch = 1000 steps of batch size = 100)  Spookynet actually completed in about half the time, but so3krates yielded better results and converged to a result better than Spookynet's best dramatically faster (at epoch 400).  The final result of so3krates was about 18% better on the force than Spookynet (energy was actually a little worse than Spookynet but I don't care about that as much).  So this is a pretty big improvement.

Now I do need to re-run my Spookynet model on the full dataset of approximately 2M molecules since that dramatically improves the error and I was using a slightly different loss function before.  So I am going to do that.  I will need to either rent some servers or upgrade my RAM here to do the equivalent test on So3krates, which uses far more memory.  In the meantime, I am going to look to adapt your code to pytorch since I assume that the memory issues etc are due to the way things are implemented and I can probably update that.   I am hoping I can replace the Spookynet layers with So3krates, although this looks like a challenge!

**Comment by shubbey** [comment]
Hey, just bumping this to see if you could offer some advice on the port.  I got ahold of a better machine with multiple gpus so I've worked my pytorch code to do distributed training; I'm pretty much confined to pytorch due to all of the infrastructure we've setup here but I am still eager to try your code.  I will see what I can do on my own but obviously you have a lot more expertise.  Thank you.
